lib/rspec/support/method_signature_verifier.rb

```diff
@@ -154,19 +154,57 @@ def classify_parameters
       INFINITY = 1 / 0.0
     end
 
-    # Some versions of JRuby have a nasty bug we have to work around :(.
-    # https://github.com/jruby/jruby/issues/2816
+    # JRuby has only partial support for UnboundMethod#parameters, so we fall back on using #arity
+    # https://github.com/jruby/jruby/issues/2816 and https://github.com/jruby/jruby/issues/2817
     if RSpec::Support::Ruby.jruby? &&
        RubyFeatures.optional_and_splat_args_supported? &&
-       Class.new { attr_writer :foo }.instance_method(:foo=).parameters == []
+       Java::JavaLang::String.instance_method(:char_at).parameters == []
+      class MethodSignature < remove_const(:MethodSignature)
+      private
+
+        def classify_parameters
+          super
+          if (arity = @method.arity) != 0 && @method.parameters.empty?
+            if arity < 0
+              @min_non_kw_args = ~arity
+              @max_non_kw_args = INFINITY
+            else
+              @min_non_kw_args = arity
+              @max_non_kw_args = arity
+            end
+          end
+        end
+      end
+    end
 
+    if RSpec::Support::Ruby.jruby? &&
+       Java::JavaLang::String.instance_method(:char_at).arity == -1
       class MethodSignature < remove_const(:MethodSignature)
       private
 
         def classify_parameters
           super
-          return unless @method.parameters == [] && @method.arity == 1
-          @max_non_kw_args = @min_non_kw_args = 1
+          if @method.arity == -1 && (arity = java_method_arity)
+            if arity < 0
+              @min_non_kw_args = ~arity
+              @max_non_kw_args = INFINITY
+            else
+              @min_non_kw_args = arity
+              @max_non_kw_args = arity
+            end
+          end
+        end
+
+        def java_method_arity
+          if @method.owner.respond_to?(:java_class)
```

Comment:
I'd prefer to use an early-exit guard, like so:

``` ruby
return unless @method.owner.respond_to?(:java_class)
```

Reply:
Fixed